File: src/Web/JavaScriptJS/example3.js
// Abstract Factory
function createReport(reportType) {
    if (reportType === "pdf") {
        return new PDFReport();
    } else if (reportType === "html") {
        return new HTMLReport();
    }
}

// Concrete Products
class PDFReport {
    generate() {
        console.log("Generating PDF report");
    }
}

class HTMLReport {
    generate() {
        console.log("Generating HTML report");
    }
}

// Usage
const report1 = createReport("pdf");
report1.generate();

const report2 = createReport("html");
report2.generate();

File: src/Niche/Prolog/example3.pl
% Abstract Factory
create_report(pdf, pdf_report).
create_report(html, html_report).

% Concrete Products
generate(pdf_report) :- write('Generating PDF report'), nl.
generate(html_report) :- write('Generating HTML report'), nl.

% Usage
run :-
    create_report(pdf, PDFReport),
    generate(PDFReport),
    create_report(html, HTMLReport),
    generate(HTMLReport).
